libraries/chain/controller.cpp

```diff
@@ -1383,10 +1439,58 @@ struct controller_impl {
       if( start_block_num <= blog_head->block_num() ) {
          ilog( "existing block log, attempting to replay from ${s} to ${n} blocks", ("s", start_block_num)("n", blog_head->block_num()) );
          try {
+            std::vector<block_state_legacy_ptr> legacy_branch; // for blocks that will need to be converted to IF blocks
             while( auto next = blog.read_block_by_num( chain_head.block_num() + 1 ) ) {
+               apply_l<void>(chain_head, [&](const auto& head) {
+                  if (next->is_proper_svnn_block()) {
+                     const bool skip_validate_signee = true; // validated already or not in replay_push_block according to conf.force_all_checks;
+                     assert(!legacy_branch.empty()); // should have started with a block_state chain_head or we transition during replay
+                     // transition to savanna
+                     block_state_ptr prev;
+                     for (size_t i = 0; i < legacy_branch.size(); ++i) {
+                        if (i == 0) {
+                           prev = block_state::create_if_genesis_block(*legacy_branch[0]);
+                        } else {
+                           const auto& bspl = legacy_branch[i];
+                           auto new_bsp = std::make_shared<block_state>(
+                                 *prev,
+                                 bspl->block,
+                                 protocol_features.get_protocol_feature_set(),
+                                 validator_t{}, skip_validate_signee);
+                           // legacy_branch is from head, all should be validated
+                           assert(bspl->action_receipt_digests);
+                           auto action_mroot = calculate_merkle(*bspl->action_receipt_digests);
```

Comment:
Same, I don't believe you'll be able to reuse the `action_receipt_digests` since they will be different for legacy and Savanna blocks.